components/brave_shields/content/browser/brave_shields_p3a.h

```diff
@@ -58,37 +58,107 @@ enum ShieldsIconUsage {
   kSize,
 };
 
+// This code implements a lazy evaluation pattern for P3A logging calls with
+// several key features:
+//
+// * Incognito Mode Handling - The P3AContextCheck ensures analytics are only
+//   recorded in non-incognito contexts by verifying IsOffTheRecord().
+//
+// * Lazy Evaluation - The P3A_LAZY_CALLS macro prevents unnecessary evaluation
+//   of logging parameters when P3A is disabled for the current context.
+//
+// * Usage Enforcement - The P3AResult class, marked with [[nodiscard]], along
+//   with its used_ flag, ensures logging calls are properly handled within
+//   the context check.
+//
+// * The P3A() macro provides a clean interface that automatically handles
+//   all context checking logic.
+
+// Only use this if you can't acquire context. Make sure to use it in
+// non-incognito mode.
+struct NotIncognito {
+  bool IsOffTheRecord() const { return false; }
+};
+
+struct P3AContextCheck {
+  template <typename Context>
+  static bool From(const Context context) {
+    if constexpr (std::is_pointer_v<Context>) {
+      if (!context || context->IsOffTheRecord()) {
+        return false;
+      }
+    } else {
+      if (context.IsOffTheRecord()) {
+        return false;
+      }
+    }
+    return true;
+  }
+};
+
+class P3AResult {
+ public:
+  ~P3AResult() { CHECK(used_); }
+
+ private:
+  friend struct P3ACallStream;
+  bool used_ = false;
+};
+
+struct P3ACallStream {
+  struct Voidify {
+    void operator&(P3ACallStream&) {}
+  };
+
+  P3ACallStream& operator<<(P3AResult&& r) {
+    r.used_ = true;
+    return *this;
+  }
+};
+
+#define P3A_LAZY_CALLS(stream, is_enabled) \
+  !(is_enabled) ? (void)0 : ::brave_shields::P3ACallStream::Voidify() & (stream)
+
+#define P3A(context)                               \
```

Comment:
very clever, but imo this may be overkill given the relatively small amount of call sites. wdyt about just passing the map into the relevant "Record" functions and performing a simple OTR check inside each function?

Reply:
I've moved this to a separate Draft PR. This PR only adds an explicit check for `RecordShieldsToggled`. 

https://github.com/brave/brave-core/pull/29383